File: python/median_finder.py
import heapq
class MedianFinder(object):

    def __init__(self):
        self.smaller = []   
        self.larger = []

    # O(m * log(n)) time | O(n) space
    def addNum(self, num):
        """
        :type num: int
        :rtype: None
        """
        if len(self.smaller) == 0:
            heapq.heappush(self.smaller, -1 * num)
        else:
            peek = -self.smaller[0]
            if num > peek:
                heapq.heappush(self.larger, num)
            else:
                heapq.heappush(self.smaller, -1 * num)

        self.rebalance()
        
    def rebalance(self):
        if len(self.smaller) > len(self.larger) + 1:
            small = heapq.heappop(self.smaller)
            heapq.heappush(self.larger, -1 * small)
        elif len(self.larger) > len(self.smaller) + 1:
            large = heapq.heappop(self.larger)
            heapq.heappush(self.smaller, -1 * large)


    # O(m) time
    def findMedian(self):
        """
        :rtype: float
        """
        if len(self.smaller) > len(self.larger):
            return -self.smaller[0]
        elif len(self.larger) > len(self.smaller):
            return self.larger[0]
        else:
            return (-self.smaller[0] + self.larger[0]) / 2.0
    
solution = MedianFinder()

solution.addNum(1)
solution.addNum(2) 

print(solution.findMedian())

solution.addNum(3)
print(solution.findMedian())

File: src/main/java/august_2025/MedianFinder.java
package august_2025;

import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.concurrent.atomic.LongAccumulator;

public class MedianFinder {

    // m - function call
    // n - length of array
    PriorityQueue<Integer> smaller;
    PriorityQueue<Integer> larger;

    public MedianFinder() {
        smaller = new PriorityQueue<>(Comparator.reverseOrder());
        larger = new PriorityQueue<>();
    }

    // O(m * log(n)) time | o(n) space
    public void addNum(int num) {
        if (smaller.isEmpty()) {
            smaller.add(num);
        } else {
            int small = smaller.peek();
            if (num > small) {
                larger.add(num);
            } else {
                smaller.add(num);
            }
        }
        rebalance();
    }

    private void rebalance() {
        if (smaller.size() - larger.size() > 1) {
            larger.add(smaller.poll());
        } else if (larger.size() - smaller.size() > 1) {
            smaller.add(larger.poll());
        }
    }

    // O(m) time | O(n) space
    public double findMedian() {
        if (smaller.size() > larger.size()) {
            return smaller.peek();
        } else if (larger.size() > smaller.size()) {
            return larger.peek();
        } else {
            return 1.0 * (smaller.peek() + larger.element()) / 2;
        }
    }

}
